elements/lisk-tree/src/sparse_merkle_tree/sparse_merkle_tree.ts

```diff
@@ -239,9 +260,171 @@ export class SparseMerkleTree {
+				currentNode = await this.getNode(currentNode.rightHash);
+			}
+
+			if (currentNodeSibling instanceof Empty) {
+				binaryBitmap = `0${binaryBitmap}`;
+			} else {
+				binaryBitmap = `1${binaryBitmap}`;
+				siblingHashes.push(currentNodeSibling.hash);
+			}
+			h += 1;
+		}
+
+		if (currentNode instanceof Empty) {
+			// exclusion proof
+			return {
+				siblingHashes,
+				ancestorHashes,
+				binaryBitmap,
+				key: queryKey,
+				value: EMPTY_VALUE,
+			};
+		}
+
+		if (
+			currentNode instanceof Leaf &&
+			currentNode.key.toString('hex') !== queryKey.toString('hex')
+		) {
+			// exclusion proof
+			ancestorHashes.push(currentNode.hash); // in case the leaf is sibling to another node
+			return {
+				siblingHashes,
+				ancestorHashes,
+				binaryBitmap,
+				key: currentNode.key,
+				value: currentNode.value,
+			};
+		}
+		if (
+			currentNode instanceof Leaf &&
+			currentNode.key.toString('hex') === queryKey.toString('hex')
+		) {
+			// inclusion proof
+			ancestorHashes.push(currentNode.hash); // in case the leaf is sibling to another node
+			return {
+				siblingHashes,
+				ancestorHashes,
+				binaryBitmap,
+				key: currentNode.key,
+				value: currentNode.value,
+			};
+		}
+		return {
+			key: queryKey,
+			value: EMPTY_VALUE,
+			binaryBitmap: bufferToBinaryString(EMPTY_HASH),
+			siblingHashes: [],
+			ancestorHashes: [],
+		};
+	};
+
+	public generateMultiProof = async (queryKeys: Buffer[]): Promise<Proof> => {
+		const partialQueries: SingleProof[] = [];
+		for (const queryKey of queryKeys) {
+			const query = await this.generateSingleProof(queryKey);
+			partialQueries.push(query);
+		}
+
+		const queries: Query[] = [...partialQueries].map(sp => ({
+			bitmap: binaryStringToBuffer(sp.binaryBitmap),
+			key: sp.key,
+			value: sp.value,
+		}));
+		const siblingHashes: Buffer[] = [];
+		const ancestorHashes = [...partialQueries].map(sp => sp.ancestorHashes).flat();
+		let sortedQueries: QueryWithHeight[] = [...partialQueries].map(sp => ({
+			binaryBitmap: sp.binaryBitmap,
+			key: sp.key,
+			value: sp.value,
+			siblingHashes: sp.siblingHashes,
+			height: sp.binaryBitmap.length,
+		}));
+		sortedQueries = sortByBitmapAndKey(sortedQueries);
+
+		while (sortedQueries.length > 0) {
+			const sp = sortedQueries.shift()!;
+			if (sp.height === 0) {
+				continue;
+			}
+			const b = sp.binaryBitmap.charAt(sp.binaryBitmap.length - sp.height);
+			if (b === '1') {
+				const nodeHash = sp.siblingHashes.pop()!;
+				let isPresentInSiblingHashes = false;
+				let isPresentInAncestorHashes = false;
+				for (const i of siblingHashes) {
+					if (i.equals(nodeHash)) isPresentInSiblingHashes = true;
+				}
+				for (const i of ancestorHashes) {
+					if (i.equals(nodeHash)) isPresentInAncestorHashes = true;
+				}
+				if (!isPresentInSiblingHashes && !isPresentInAncestorHashes) {
+					// TODO : optimize this
+					siblingHashes.push(nodeHash);
+				}
+			}
+			sp.height -= 1;
+
+			if (sortedQueries.length > 0) {
```

Comment:
where is this logic matching to in the LIP?

Reply:
In the LIP it is mentioned in the beginning of the `while` loop :
```
        sortedQueries = sortByBitmapAndKey(sortedQueries)
```